js/__internal/viz/core/m_base_widget.ts

```diff
@@ -0,0 +1,768 @@
+import _Layout from '@js/viz/core/layout';
+import { Renderer } from '@js/viz/core/renderers/renderer';
+import { parseScalar as _parseScalar } from '@js/viz/core/utils';
+
+import graphicObject from '../../common/m_charts';
+
+const { floor } = Math;
+const { log } = warnings;
+const SIZE_CHANGING_THRESHOLD = 1;
+
+const OPTION_RTL_ENABLED = 'rtlEnabled';
+
+const SIZED_ELEMENT_CLASS = 'dx-sized-element';
+
+const baseOptionMethod = DOMComponent.prototype.option;
+
+function getTrue(): boolean {
+  return true;
+}
+
+function getFalse(): boolean {
+  return false;
+}
+
+function areCanvasesDifferent(canvas1, canvas2): boolean {
+  return !(Math.abs(canvas1.width - canvas2.width) < SIZE_CHANGING_THRESHOLD && Math.abs(canvas1.height - canvas2.height) < SIZE_CHANGING_THRESHOLD
+        && canvas1.left === canvas2.left && canvas1.top === canvas2.top && canvas1.right === canvas2.right && canvas1.bottom === canvas2.bottom);
+}
+
+function defaultOnIncidentOccurred(e): void {
+  if (!e.component._eventsStrategy.hasEvent('incidentOccurred')) {
+    log.apply(null, [e.target.id].concat(e.target.args || []));
+  }
+}
+
+function pickPositiveValue(values): number {
+  return values.reduce((result, value) => (value > 0 && !result ? value : result), 0);
+}
+
+// TODO - Changes handling
+// * Provide more validation - something like
+//     _changes: [{
+//         code: "THEME",
+//         options: ["theme"],
+//         type: "option",
+//         handler: function () {
+//             this._setThemeAndRtl();
+//         }
+//     }, {
+//         code: "CONTAINER_SIZE",
+//         options: ["size", "option"],
+//         type: "layout",
+//         handler: function () {
+//             this._updateSize();
+//         }
+//     }]
+
+const getEmptyComponent = function () {
+  const emptyComponentConfig = {
+    _initTemplates() {},
+    ctor(element, options) {
+      this.callBase(element, options);
+      const sizedElement = domAdapter.createElement('div');
+
+      const width = options && isNumeric(options.width) ? `${options.width}px` : '100%';
+      const height = options && isNumeric(options.height) ? `${options.height}px` : `${this._getDefaultSize().height}px`;
+
+      domAdapter.setStyle(sizedElement, 'width', width);
+      domAdapter.setStyle(sizedElement, 'height', height);
+
+      domAdapter.setClass(sizedElement, SIZED_ELEMENT_CLASS, false);
+      domAdapter.insertElement(element, sizedElement);
+    },
+  };
+
+  const EmptyComponent = (DOMComponent as any).inherit(emptyComponentConfig);
+  const originalInherit = EmptyComponent.inherit;
+
+  EmptyComponent.inherit = function (config) {
+    Object.keys(config).forEach((field) => {
+      if (isFunction(config[field]) && field.substr(0, 1) !== '_' && field !== 'option' || field === '_dispose' || field === '_optionChanged') {
+        config[field] = noop;
+      }
+    });
+
+    return originalInherit.call(this, config);
+  };
+
+  return EmptyComponent;
+};
```

Comment:
cannot we move it to a component?

Reply:
Do you mean, move this component to the separated file, as separated component?